Masterbar: add Tracks instrumentation

rodrigoi opened the pull request:
Fixes #7738
~~Depends on D7169-code~~

#### Changes proposed in this Pull Request:

~~Enqueues tracks scripts when the WordPress.com masterbar is loaded, to make them available to the javascript overrides.~~

Makes the Masterbar and the Notifications Panel interactions trackable by making the record event function global on `tracks-ajax.js` and adding event handling to the Masterbar plugin.

#### Testing instructions:

* ~~Sandbox `s0.wp.com`.~~
* ~~Apply D7169-code on your sandbox.~~
* ~~Using a http debugger, click on any menu item (this does not work on top level items). An entry for the corresponding event with the domain `pixel.wp.com` should appear.~~

* Run a WordPress instance with this branch
* Enable WordPress.com Masterbar
* Using an http debugger, any activity on the masterbar, or opening a site, post or comment on the notifications panel should trigger an `admin-ajax.php` request with the `jetpack_tracks` action.

vindl commented on modules/masterbar/masterbar.php:
```diff
@@ -123,6 +123,14 @@ public function add_styles_and_scripts() {
+
+		// so tracks events can be fired from the frontend.
+		wp_enqueue_script( 'jp-tracks', '//stats.wp.com/w.js', array( 'a8c_wpcom_masterbar_overrides' ), gmdate( 'YW' ), true );
+		wp_enqueue_script( 'jp-tracks-functions', plugins_url( '_inc/lib/tracks/tracks-callables.js', JETPACK__PLUGIN_FILE ), array( 'a8c_wpcom_masterbar_overrides' ), JETPACK__VERSION, false );
```
The last argument (`$in_footer`) to `wp_enqueue_script` is `false` by default so we don't need to include it here.
https://developer.wordpress.org/reference/functions/wp_enqueue_script/

vindl replied:
One more thing occurred to me - shouldn't the dependencies be the other way around? Since we are using the tracks code in `masterbar.js`, should we instead add `jp-tracks` and `jp-tracks-functions` as dependencies for `a8c_wpcom_masterbar_overrides`?

rodrigoi commented:
@vindl thanks for your comments!. I've done some refactoring on the overrides file that affects this PR.

- Only enqueue `stats.wp.com` JS. We no longer add the `tracks-callable.js` file.
- The configuration now lives in JP.
- We no longer set variables on `masterbar-overrides/masterbar.js`. 

I think this is ready for another review 😉

dereksmart commented:
JSHint is failing for tracks-events.js, can you resolve? 

You can run `gulp js:hint` to test. 

```modules/masterbar/tracks-events.js
  line 54  col 88   Extra comma. (it breaks older versions of IE)
  line 60  col 97   Missing semicolon.
  line 64  col 97   Missing semicolon.
  line 68  col 125  Missing semicolon.
  line 69  col 10   Extra comma. (it breaks older versions of IE)
  line 72  col 38   'jetpackTracks' is not defined.
  line 73  col 5    'jetpackTracks' is not defined.```

rodrigoi commented:
@Automattic/lannister @Automattic/jetpack I've made some changes I'd like y'all to check. The first change is the addition of `tracks-events.js` to the plugin. That file includes the code on D7169-code, and the idea is to make deployment of these changes easier by no longer depending on a file hosted on WordPress.com.

The second change is the addition of a very rough (VERY ROUGH) version of D7236-code that handles a new `jetpack-track-and-bounce.php` route. To add some level of security, the requests to that route have a _nonce_, and the target of the link is matched against a whitelist on the client and a dictionary on the server.
I'm feeling more comfortable having that redirect handled internally on Jetpack, at least until the universal solution on D7236-code is complete.
If that's a direction we'll like to pursue, my intention is to move that new code to `class.jetpack-tracks.php`. I'm eagerly awaiting your comments about this. I insist, it's very rough and needs polishing, so, any comment is welcome.

jeremysawesome commented on modules/masterbar/tracks-events.js:
```diff
@@ -1,60 +1,35 @@
-(function() {
-	window.wpcom_masterbar = window.wpcom_masterbar || {};
+(function( $, jetpackTracks ) {
+	window._tkq = window._tkq || [];
```
Thoughts on injecting `window` into the IIFE as well?

jeremysawesome commented on modules/masterbar/tracks-events.js:
```diff
@@ -69,5 +44,65 @@
+		window._tkq.push( [ 'recordEvent', eventData.name, eventData.properties( data ) ] );
+	} );
+
+	window.jetpackTracks = null;
```
why set this to null here?

rodrigoi replied:
the intention is to remove the _nonce_ from the global scope so it can't be used by other functions.

jeremysawesome commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +118,50 @@ function __construct() {
+	}
+
+	private function tracks_and_redirect_get_target( $trackId ) {
+		return "http://wordpress.com" . str_replace("{blog_url}", Jetpack_Options::get_option( 'id' ),  $this->links_tracks_events[ $trackId ][1] );
```
I think we should be able to use single quotes here around `http://wordpress.com` and `{blog_url}`. Also, maybe consider using `https` instead of `http` in the wordpress.com url.

vindl replied:
Agreed, we should use `https` instead. Also we can replace `Jetpack_Options::get_option( 'id' )` with the existing `$this->primary_site_url`.

Also would it make sense to simplify this and avoid `str_replace` completely. We can initialize `links_tracks_events` in constructor after `primary_site_url` is set, so we don't have to use `{blog_url}` placeholder.

rodrigoi replied:
Sorry for the typo. There's no emoji for shame...

I've updated the code to initialize the tracks events on the constructor and remove the `tracks_and_redirect_get_target ` altogether on a36123157b1f1350f06c11dc27d73f714c68a981.

jeremysawesome commented on modules/masterbar/masterbar.php:
```diff
@@ -22,6 +22,52 @@ class A8C_WPCOM_Masterbar {
+		//my sites - configure
+		"wp-admin-bar-sharing"                    => array( "masterbar_configure_sharing_link_click", "/sharing/{blog_url}" ),
+		"wp-admin-bar-users-toolbar"              => array( "masterbar_configure_people_link_click", "/people/team/{blog_url}" ),
+		"wp-admin-bar-users-toolbar-secondary"    => array( "masterbar_configure_people_add_button_click", "/wp-admin/user-new.php" ),
```
Should we use `/people/new/{blog_url}` here instead of `/wp-admin/user-new.php`?

rodrigoi replied:
![that is correct](https://user-images.githubusercontent.com/233601/30736631-6f3a34aa-9f7b-11e7-941b-a09f333c4588.gif)

Updated on f24b321101188d2bfbfe80d8e8b8fb10e49461d2.

jeremysawesome commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +118,50 @@ function __construct() {
+		return $query_vars;
+	}
+
+	public function tracks_and_redirect_parse_request( &$wp ) {
```
Do we need to pass by reference here?

rodrigoi replied:
Correct again! Updated on a36123157b1f1350f06c11dc27d73f714c68a981.

jeremysawesome commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +118,50 @@ function __construct() {
+
+			JetpackTracking::record_user_event(
+				$this->links_tracks_events[ $_GET[ 'tracks_and_bounce' ] ][0],
+				$data = array( 'target' => $target )
```
What are your thoughts on dropping the `$data = ` assignment and passing in the array?

rodrigoi replied:
Dropped on a36123157b1f1350f06c11dc27d73f714c68a981. Copypasta error.

jeremysawesome commented:
This is looking good to me! I like your approach 👍

vindl commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +118,50 @@ function __construct() {
+
+	public function tracks_and_redirect_parse_request( &$wp ) {
+		if (
+			array_key_exists( 'tracks_and_bounce', $wp->query_vars ) &&
```
Could we use `get_query_var` instead of accessing `$_GET` directly?

rodrigoi replied:
`get_query_var` is not quite working for this scenario... will look into it later. For the moment I've replaced it with `$wp->query_vars[...]`.

vindl replied:
Might be because the new values that we are using were not whitelisted before? (`query_vars` filter)

vindl commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +118,50 @@ function __construct() {
+				$this->links_tracks_events[ $_GET[ 'tracks_and_bounce' ] ][0],
+				$data = array( 'target' => $target )
+			);
+			wp_redirect( $target );
```
I know that we are atm building this link and it looks safe atm, but could we switch this to `wp_safe_redirect` to avoid accidentally exposing open redirect vulnerability in the future?

rodrigoi replied:
Excellent point. Updated on 3f119d6f2195812eb97a8c8343d1b5a7c18e61b1

dmsnell commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +118,50 @@ function __construct() {
 	}
 
+	public function tracks_and_redirect_add_route() {
+		add_rewrite_rule( 'jetpack-track-and-bounce.php\?tracks_and_bounce=(\w*)', 'index.php?tracks_and_bounce=$matches[1]', 'top' );
```
we may not want a star here, but a `+`

what does it mean to redirect `jetpack-track-and-bounce.php?tracks_and_bounce=`

rodrigoi commented:
@vindl we want to add tracking on WordPress.com. But I'm ok with some duplication as long as it moves things forward. And we can remove said duplication later when the universal redirect tool is deployed.

vindl commented on modules/masterbar/tracks-events.js:
```diff
@@ -0,0 +1,109 @@
+	var notesTracksEvents = {
+		openSite: {
+			name: 'jetpack_masterbar_notifications_open_site',
+			properties: function( data ) { return { site_id: data.siteId, post_id: data.postId }; }
```
This might also be more readable if move the function body to a new line (same goes for two cases bellow).

rodrigoi replied:
Done. Changed on d2d94d956eb86c0266662e559e68f92a69ba679c

vindl commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +76,104 @@ function __construct() {
+
+		JetpackTracking::record_user_event(
+			$track_event[0],
+			array( 'target' => $track_event[1] )
```
What's the benefit of recording `target` here? I don't expect these to change, and even if they do it seems to me that we are only interested in recording user's action of navigating to particular page, and not in the corresponding URL.

rodrigoi replied:
Having a single tracks event now, it make sense to store the target.

vindl commented on modules/masterbar/tracks-events.js:
```diff
@@ -0,0 +1,109 @@
+		}
+	}
+
+	$( document ).ready( function(){
```
Missing space before `{`.

rodrigoi replied:
Fixed :)

vindl commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +76,104 @@ function __construct() {
+		return $query_vars;
+	}
+
+	public function tracks_and_redirect_parse_request( $query ) {
```
Would it be possible to move code that is not specific to Masterbar to a separate class so it can be reused elsewhere if needed?

rodrigoi replied:
I've removed all the tracks specific code from the Masterbar into the `JetpackTracking` class.

vindl commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +76,104 @@ function __construct() {
+		}
+
+		$trackId = $query->query_vars[ 'tracks_and_bounce' ];
+		if ( ! array_key_exists( $trackId, $this->tracks_events ) ) {
```
Just wanted to note that there is an implicit level of trust here, since Jetpack site owners can modify the whitelist, which would allow them to record bogus events unless we validate them on dotcom side.

dereksmart commented:
@rodrigoi code freeze is end of day tomorrow, any chance you can tackle the requests?

gwwar commented:
Thanks for all of the reviews!

@dereksmart he's out until this Thursday, so it's unlikely we'll be able to make it. Let's aim for the release after this one.

jeremysawesome commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +76,104 @@ function __construct() {
+	}
+
+	public function tracks_and_redirect_add_route() {
+		add_rewrite_rule( 'jetpack-track-and-bounce.php\?tracks_and_bounce=(\w*)', 'index.php?tracks_and_bounce=$matches[1]', 'top' );
```
Question - what is the rewrite rule helping us to accomplish here? Is it simply to allow us to have a more descriptive and explicit name for our urls?

rodrigoi replied:
yeah, basically the idea is to have an explicit address, but It's doing more harm than good. I'll replace it...

jeremysawesome replied:
If you think it needs to be replaced, go for it 👍  
I was just trying to understand what it was for 😄

rodrigoi commented on modules/masterbar/tracks-events.js:
```diff
@@ -0,0 +1,109 @@
+
+	var linksWhitelist = [
+		//my sites - top items
+		'wp-admin-bar-switch-site', 'wp-admin-bar-blog-info', 'wp-admin-bar-site-view', 'wp-admin-bar-blog-stats', 'wp-admin-bar-plan', 'wp-admin-bar-plan-secondary',
```
fixed. Also renamed the variable to make the intent a little more clear.

rodrigoi commented on modules/masterbar/masterbar.php:
```diff
@@ -122,7 +228,15 @@ public function add_styles_and_scripts() {
 		wp_enqueue_script( 'a8c_wpcom_masterbar_overrides', $this->wpcom_static_url( '/wp-content/mu-plugins/admin-bar/masterbar-overrides/masterbar.js' ), array( 'jquery' ), JETPACK__VERSION );
+
+		// so tracks events can be fired from the frontend.
+		wp_enqueue_script( 'jp-tracks', '//stats.wp.com/w.js', array(), gmdate( 'YW' ), true );
```
I've removed the protocol just in case the site is not loading over HTTPS. But I've removed the line anyway, because it was redundant.

rodrigoi commented on modules/masterbar/masterbar.php:
```diff
@@ -68,6 +76,104 @@ function __construct() {
+	private function initialize_tracks_events( $site_slug ) {
+		$this->tracks_events = array(
+			//my sites - top level items
+			'wp-admin-bar-switch-site'                => array( 'masterbar_switch_site_link_click', 'https://wordpress.com/sites' ),
```
we already add `jetpack_` on `class.jetpack-tracks`

https://github.com/Automattic/jetpack/blob/master/class.jetpack-tracks.php#L76

rodrigoi commented:
@dereksmart @gwwar @vindl, @withinboredom  and @jeremysawesome:  thanks for the reviews and sorry for the delay. I've very quickly:

- moved the Tracks specific code from the Masterbar into `class.jetpack-tracks.php`.
- links on the redirect mapping will record an `jetpack_track_and_bounce` event with the source (`$trackId`) and the target (mapped url). Top level items on the Masterbar and the Notifications panel still trigger their own events.
- removed the rewrite rule because it was confusing. The script just uses `index.php` now, with the query params.
- applied the suggested legibility changes.

I'll keep testing and updating the code. I still have a few hours before my flight but I'll defer to you guys for weather or not merging this into `5.4`.

rodrigoi commented:
@jeremysawesome how would you feel if we include an action you can hook into to perform other logging, like MC stats bump?

I'd like to keep the code _tracks centric_. I'm all ears if there's a better way to do it.

jeremysawesome commented:
> how would you feel if we include an action you can hook into to perform other logging, like MC stats bump?

@rodrigoi - this sounds like it could work to me. I'm not sure how _important_ adding an MC bump is. Could be that for Jetpack we don't need one. Adding an action sounds like a good way to keep the code tracks-centric while at the same time being flexible.

rodrigoi commented:
@withinboredom I just updated the code to use a single event name for all user actions, along with some other changes.

To fix the permalink issue, I've started a discussion here: p1HpG7-4fg-p2

rodrigoi commented:
@Automattic/lannister @Automattic/jetpack after the discussion on p1HpG7-4fg-p2 I went back to the original implementation, but instead of using a new `$.ajax`, I've modified `jpTracksAJAX` by adding a global `record_ajax_event` that returns a promise that can be resolved (for links) or ignored (for everything else). No more redirect shenanigans.

Let me know what you think 😉

kwight commented:
I'm not sure I understand what I'm looking for when I test this. I've applied D7169 and sandboxed `s0.wp.com`, and on my JP sandbox, I've got what I'm pretty sure is the masterbar (I can see a `.jetpack-masterbar` class in the body).

![screen shot 2017-10-05 at 11 17 25 am](https://user-images.githubusercontent.com/349751/31245484-5d7d7f26-a9bf-11e7-91f5-1bfcd79617dd.png)

Looking at D7169, it seems `window.wpcom_masterbar` should be there, which I don't have; can you give any more details of what I should be looking for or how to test this?

rodrigoi commented:
@withinboredom would you mind giving this another try 😉

rodrigoi commented on _inc/lib/tracks/tracks-ajax.js:
```diff
@@ -1,26 +1,37 @@
 /* global jpTracksAJAX, jQuery */
 
 (function( $, jpTracksAJAX ) {
+	window.jpTracksAJAX = window.jpTracksAJAX || {};
```
It's just me being super explicit that we'll be using and modifying a global object, that can't be `null` or `undefined`, and preventing any problem we may have when loading the script.

withinboredom commented on modules/masterbar/masterbar.php:
```diff
@@ -21,6 +21,7 @@ class A8C_WPCOM_Masterbar {
 	private $primary_site_slug;
 	private $user_text_direction;
 	private $user_site_count;
+	private $tracks_events;
```
Can we remove this if it's no longer needed?

rodrigoi replied:
great catch! Removed on 474d8cef923a9f8f839d733033a496d0a3a28089.
